src/viam/examples/modules/mlmodelservice_tflite.cpp

```diff
@@ -0,0 +1,759 @@
+#include <pthread.h>
+#include <signal.h>
+
+#include <fstream>
+#include <iostream>
+#include <mutex>
+#include <sstream>
+#include <stdexcept>
+
+#include <grpcpp/channel.h>
+#include <grpcpp/create_channel.h>
+#include <grpcpp/security/credentials.h>
+#include <tensorflow/lite/c/c_api.h>
+
+#include <viam/sdk/components/component.hpp>
+#include <viam/sdk/config/resource.hpp>
+#include <viam/sdk/module/service.hpp>
+#include <viam/sdk/registry/registry.hpp>
+#include <viam/sdk/rpc/server.hpp>
+#include <viam/sdk/services/mlmodel/mlmodel.hpp>
+#include <viam/sdk/services/mlmodel/server.hpp>
+
+namespace {
+
+namespace vsdk = ::viam::sdk;
+constexpr char service_name[] = "mlmodelservice_tflite";
+
+class MLModelServiceTFLite : public vsdk::MLModelService {
+    class write_to_tflite_tensor_visitor_;
+
+   public:
+    explicit MLModelServiceTFLite(vsdk::Dependencies dependencies,
+                                  vsdk::ResourceConfig configuration)
+        : MLModelService(configuration.name()),
+          state_(reconfigure_(std::move(dependencies), std::move(configuration))) {}
+
+    ~MLModelServiceTFLite() final {
+        // All invocations arrive via gRPC, so we know we are idle
+        // here. It should be safe to tear down all state
+        // automatically without needing to wait for anything more to
+        // drain.
+    }
+
+    void reconfigure(vsdk::Dependencies dependencies, vsdk::ResourceConfig configuration) final {
+        // Care needs to be taken during reconfiguration. There may
+        // not be higher level protection against invocation during
+        // reconfiguration. Keep all state in a shared_ptr managed
+        // block, and allow client invocations to act against current
+        // state while a new configuration is built, then swap in the
+        // new state. State which is in use by existing invocations
+        // will remain valid until the clients drain. If
+        // reconfiguration fails, fall back to the old state, though
+        // in practice the component is likely to get restarted.
+
+        // Swap out the state_ member with nullptr. Existing infocations
+        using std::swap;
+        std::shared_ptr<state> state;
+        {
+            // Wait until we have a state in play, then take
+            // ownership, so that we don't race with other
+            // reconfigurations and so other invocations wait on a new
+            // state.
+            std::unique_lock<std::mutex> lock(state_lock_);
+            state_ready_.wait(lock, [this]() { return state_ != nullptr; });
+            swap(state_, state);
+        }
+        try {
+            state = reconfigure_(std::move(dependencies), std::move(configuration));
+        } catch (...) {
+            // If we failed to reconfigure, restore the original state and unblock waiters.
```

Comment:
If we failed to reconfigure, the resource should behave as if it is closed and not allow requests.

Reply:
Great. I wasn't sure what was expected. I think I can make this somewhat simpler if I can allow the service to become closed on a failed reconfiguration.